gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2

```diff
@@ -0,0 +1,200 @@
+
+    @property
+    def operations_client(self) -> operations_v1.OperationsClient:
+        """Create the client designed to process long-running operations.
+
+        This property caches on the instance; repeated calls return the same
+        client.
+        """
+        # Sanity check: Only create a new client if we do not already have one.
+        if 'operations_client' not in self.__dict__:
+            from google.api_core import grpc_helpers
+            self.__dict__['operations_client'] = operations_v1.OperationsClient(
+                grpc_helpers.create_channel(
+                    self._host,
+                    credentials=self._credentials,
+                    scopes=self.AUTH_SCOPES,
+                )
+            )
+
+        # Return the client from cache.
+        return self.__dict__['operations_client']
+    {%- endif %}
+
+
+    {%- for method in service.methods.values() %}
+    {# TODO(yonmg): consider using enums #}
+        {%- set ns = namespace(http=None) %}
+        {%- for option in method.options.ListFields() %}
+            {%- if option[0].name == 'http' %}
+                {%- set ns.http = option[1] %}
+            {%- endif %}
+        {%- endfor %}
+
+        {%- if ns.http %}
+
+    
+    def {{ method.name|snake_case }}(self,
+            request: {{ method.input.ident }}, *,
+            metadata: Sequence[Tuple[str, str]] = (),
+            ) -> {{ method.output.ident }}:
+        r"""Call the {{- ' ' -}}
+        {{ (method.name|snake_case).replace('_',' ')|wrap(
+                width=70, offset=45, indent=8) }}
+        {{- ' ' -}} method over HTTP.
+
+        Args:
+            request (~.{{ method.input.ident }}):
+                The request object.
+                {{ method.input.meta.doc|rst(width=72, indent=16) }}
+            metadata (Sequence[Tuple[str, str]]): Strings which should be
+                sent along with the request as metadata.
+        {%- if not method.void %}
+
+        Returns:
+            ~.{{ method.output.ident }}:
+                {{ method.output.meta.doc|rst(width=72, indent=16) }}
+        {%- endif %}
+        """
+
+        # Check which http method to use
+        {%- if ns.http.get != '' %}
+            {%- set ns.httpMethod = {'method':'get','url':ns.http.get} %}
+        {%- elif ns.http.put != '' %}
+            {%- set ns.httpMethod = {'method':'put','url':ns.http.put} %}
+        {%- elif ns.http.post != '' %}
+            {%- set ns.httpMethod = {'method':'post','url':ns.http.post} %}
+        {%- elif ns.http.delete != '' %}
+            {%- set ns.httpMethod = {'method':'delete','url':ns.http.delete} %}
+        {%- elif ns.http.patch != '' %}
+            {%- set ns.httpMethod = {'method':'patch','url':ns.http.patch} %}
+        {%- elif ns.http.custom != '' %}
+            {%- set ns.httpMethod = {'method':'custom','url':ns.http.custom} %}
+        {%- else %}
+            {%- set ns.httpMethod = None %}
+        {%- endif %}
```

Comment:
I'm leaning towards anything that involves state modification living in the generator datatypes. It's easier to test and easier to change if we need to.

Reply:
I'm not sure I understand what you mean by generator datatypes. Could you clarify?

Reply:
The `gapic/schema/wrappers.py:{MessageType,Service,Method}` classes. The auxiliary datastructures that the generate creates before iterating over the templates.